Dockerfile

```diff
@@ -16,7 +16,18 @@ RUN YARN_VERSION=1.17.3 \
   && rm yarn-v$YARN_VERSION.tar.gz
 
 COPY requirements.txt $INSTALL_PATH/requirements.txt
-RUN apt-get install -qq -y python python-pip python-setuptools python-wheel --fix-missing --no-install-recommends && pip install --quiet -r requirements.txt
+# This should be kept in sync with the version specified in runtime.txt
```

Comment:
Is there any way to have Dockerfile load environment variables from another file, so we don't have multiple copies of this setting to keep in sync?

Reply:
To my knowledge, not directly. I think we'd have to `COPY` the `runtime.txt` file into the container, and then add a part to the `RUN` which parses this file and extracts the version number. As I mentioned in a commit message, I didn't feel it was worth spending time on this given how infrequently we are likely to update Python.

Reply:
(@tahb may have something more useful to say about this)

Reply:
Declaring all of our versions in one place I think is where we are aiming to get to, and we are moving towards that being our Dockerfile.

Given we are not using Docker in live environments at the moment but the testing I am doing to make this change is going well, I suggest that this is proposal is good enough until we can remove the dependency on `runtime.txt` itself. We would then have to amend this line `pip3 install --quiet -r $INSTALL_PATH/requirements.txt` to remove reference to it.

[As mentioned on Slack](https://dxw.slack.com/archives/CD3MUQ5TQ/p1565863818014600) you could go a step further to try and optimise this installation by moving it to in a multi-stage build. That gives us a few benefits:

* cacheable, decreasing build times
* ensures we only include the files we want in the runnable image - no compressed file or tools like `make` are left in by accident 
* smaller images take less time to move over the internet, and with less packages have less surface area for vulnerabilities

Saying that I'm not sure how well python will react if you try and move it without the compilers etc but I did find someone using `wheel` (whatever that is) to achieve it: https://www.blogfoobar.com/post/2018/02/10/python-and-docker-multistage-build

I would be happy to merge this and with or without the Docker optimisation.

Reply:
I think that since the Python change is incidental to the main change here, which is the fix for the type inference bug, I probably shouldn't spend too much time trying to understand Python well enough to be able to optimise it, _here_.

Currently, the only negative impact of this change is that it increases the build time for the development Docker image, which I think we can accept for now since we don't rebuild the images very often. As we progress with using Docker for production, we can work on optimising this. Does that sound alright?

Reply:
Sounds good to me  🙂